Use activate and deactivate conda actions in Windows.  Move activate scripts back to Conda.

msarahan opened the pull request:
This integrates https://github.com/conda/conda/pull/1707 with some minor adaptation:

- add Library/bin to PATH on windows

UPDATE:  this ended up being a little more work than anticipated.  This PR includes extensive refactoring to deal with Windows idiosyncrasies (slashes in particular).  The setting of PS1/PROMPT is now done similarly to PATH (logic handled in Python, rather than in shell scripts.)  This makes Windows and *nix consistent.  The CONDA_DEFAULT_ENV  variable has been renamed to CONDA_ACTIVE_ENV.

This PR creates new issues and potential for improvement:
  - Adding other shells should now be much simpler
  - Windows activation/deactivation is a great deal slower.  This needs investigation.

msarahan commented:
Note: slash in Library/bin is going unix-way, not Windows-way.
Tests for activate are all failing, but seems to work OK interactively.  Will debug tomorrow.

msarahan commented:
Seeing lots of messages like: 

    a = 'Exception TypeError: "\'NoneType\' object is not callable" in <function _removeHandlerRef at 0x7f1e3073e5f0> ignored

This is both on Travis and on local system.  Local system is pretty slow to activate using this new script compared to the old one.  I think it's still perfectly acceptable, but much less snappy.

msarahan commented:
activation tests are all failing on windows for me right now.  Tests seem to pass OK mostly on Travis.  Probably something about the Library/bin stuff I added.

ukoethe commented on conda/cli/activate.py:
```diff
@@ -32,11 +35,27 @@ def prefix_from_arg(arg):
+    if sys.platform == "win32":
+        path = [prefix_from_arg(arg),
+                join(prefix_from_arg(arg), 'Scripts'),
+                join(prefix_from_arg(arg), 'Library/bin'),
```
This should be 
```python
               join(prefix_from_arg(arg), join('Library', 'bin')),
```

ukoethe replied:
The previous fix was not good enough, because `Library\bin` may not even exist in some environments. A better fix is 
```python
        path = [k for k in [prefix_from_arg(arg),
                            join(prefix_from_arg(arg), 'Scripts'),
                            join(prefix_from_arg(arg), join('Library','bin')),
                           ]
                  if os.path.isdir(k)
                ]
```

msarahan replied:
This is cleaner, definitely, but I think this would pass as harmless without the loop?  If dirs on PATH don't exist, then surely nothing is found in them?

Are there potential problems with having non-existent folders on PATH that I'm missing?

ukoethe replied:
`activate.py` (line 47 below) exists with an error in this case.

msarahan replied:
well, that's a pretty good reason to fix it. =)

ukoethe commented on conda/cli/activate.py:
```diff
@@ -32,11 +35,27 @@ def prefix_from_arg(arg):
+        path = [join(prefix_from_arg(arg), 'bin')]
+    for p in path:
+        if not isdir(p):
+            sys.exit("Error: no such directory: %s" % p)
```
This test is redundant after my fix. It's also not clear to me if exiting is the desired behavior here.

msarahan replied:
Removed, thanks.

ukoethe commented on conda/cli/activate.py:
```diff
@@ -62,7 +81,7 @@ def main():
             raise
         paths = []
-        sys.stderr.write("discarding %s from PATH\n" % binpath)
+        sys.stderr.write("discarding %s from PATH\n" % pathlist_to_str(binpath))
```
In contrast to the old design, this command also removes the root environment's `Scripts` directory. Since `activate.bat` and `deactivate.bat` only exist in the root environment (because `conda` can only reside there), these commands are no longer available after the first `activate some_env` call. The error was initially masked because `conda ..deactivate` never succeeded in the root environment due to the missing `Library\bin` subdirectory. 

The upshot is that the root environment's `Scripts` directory must always remain in the PATH, but it must come after the `Scripts` directory of the active environment.

msarahan replied:
I would like to keep it so that things are functional with NO conda/miniconda/anaconda path on PATH.

My concept for this is based around shortcuts to activate the root environment, much the same as Visual Studio does things.

This means that activate.bat from a raw terminal with no modified PATH will not work.  I think that is OK - people making the choice to not alter PATH is a reasonably advanced decision.

Do you have an opinion on this?

All this said, we need to be careful to keep the root environment on PATH, if the user has it on PATH to begin with.

ukoethe replied:
I see your point. What kind of shortcuts do you have in mind?

So far, I thought that putting `<conda_root>\Scripts` in the PATH to begin with would be the standard setup on Windows. Maybe you should do it like `git` who created another subdirectory (say, `<conda_root>\cmd`) that only contains `activate.bat`, `deactivate.bat` and the conda scripts, and that can be safely added to and kept in the PATH? Alternatively, conda could remember if `<conda_root>\Scripts` was already in the PATH and act accordingly.

msarahan replied:
I have a working prototype here: https://github.com/conda/conda-recipes/pull/439 (depends on quite recent iterations of menuinst, https://github.com/ContinuumIO/menuinst and conda is tied in there a little bit too: https://github.com/conda/conda/pull/1721/files)

General idea is to hard-code the path to activate.bat into the shortcut.  There could be more than one activate.bat, or always just one (in root).  The latter is the current implementation, I think.

Feedback welcome, of course.

ukoethe replied:
Neat.

ukoethe replied:
After executing `conda install` with the new design for the first time, I realize that `conda.bat`, `activate.bat` and `deactivate.bat` are now placed into _every_ environment. Then it is indeed no longer necessary to always keep `<conda_root>\Scripts` in the PATH.

msarahan replied:
Yes, I think this is going to be a nice improvement.  This happens in Conda's symlink step during installation: https://github.com/msarahan/conda/blob/wulmer-fix/activate-behaviour-on-windows/conda/install.py#L430

ukoethe commented:
Neat.

ukoethe commented on conda/cli/activate.py:
```diff
@@ -110,7 +134,7 @@ def main():
 
     for path in os.getenv('PATH').split(os.pathsep):
-        if path != binpath:
+        if path not in binpath:
```
I found another problem here: for some reason `activate` adds the path as `C:\...` (capital drive letter), whereas `deactivate` tries to remove `c:\...` (lower case drive letter). So, no match is found and no path is actually removed. Since the Windows file system is (still, IIRC) case-insensitive, a case-insensitive comparison should be used on Windows here.

msarahan replied:
Good catch.  That one deserves a regression test.

msarahan commented:
Update: tests are updated quite a lot.  I am currently encountering an issue with logging.  The tests depend on stdout, so this is killing the tests.

Example exception output:

    Exception TypeError: "'NoneType' object is not callable" in <function _removeHandlerRef at 0x0000000003070CF8> ignored

This might be because this is a source installation?  I have been unable to isolate the issue so far - people have had the issue elsewhere, but no real fix I can find.  Just "logger wasn't being initialized correctly."  Since this code changes nothing about the logger, and moreover changes no code calling the logger, I am at a loss.

I will try to build this as a conda package, and install it that way to see if it helps.

msarahan commented on conda/cli/activate.py:
```diff
@@ -3,14 +3,25 @@
+import errno
 
 from conda.cli.common import find_prefix_name
+import conda.config
```
Moving these imports seems to be what is causing the horrible logging messages.  I have moved them back below.

This really needs a real fix.  Local imports to fix errors are a sign of a deeper problem.

msarahan commented:
All but test_PS1 pass on my Windows machine currently .  OS X has a strange problem where the PATH gets rearranged and fails tests.  I have not been able to figure it out yet.

The issue is that I have a path to sys.prefix on PATH:

     PATH="<prefix>:other_stuff

When the activate script runs, it seems to move <prefix> to the the end of PATH:

     PATH="other_stuff:<prefix>

This rightfully makes the tests fail.  I have added debug echos of the path coming out of conda ..activate <prefix> in the activate script, and these come out as expected, but the tests then fail because stdout includes too much information

Printing os.getenv("PATH") at any stage of the process returns the correct PATH (prefix at the start, or in the middle, rather than the end).  I'm really lost on this one, and would appreciate pointers if anyone has insight.

msarahan commented:
@wulmer @ukoethe : I would appreciate your approval on this PR.  I did a lot of refactoring, and I want to make sure that it still does what you intended.

@groutr @ilanschnell @csoja , this PR is urgently necessary to make the Windows menu changes work.  We need a new Conda release before the Anaconda release.  I am building installers with these changes included from my branch, for testing purposes.  Also note: I have moved the activate and deactivate scripts back from conda-env.  We will need a new conda-env release as well.  I don't know how much more testing and fixing is necessary on that project currently.  I think it can wait until after the Anaconda release, but we need to be sure that Conda's activate and deactivate scripts overwrite the conda-env ones, or else we'll get broken behavior.

ilanschnell commented on .travis.yml:
```diff
@@ -2,7 +2,6 @@ language: python
   # We don't actually use the system Python but this keeps it organized.
   - "2.7"
-  - "3.3"
   - "3.4"
```
Conda still supports Python 3.3

ilanschnell commented:
I'm all for moving the activate scripts back into conda!  However, I'm not sure if this is the right time to do it.  Let's talk about this (maybe with @csoja) tomorrow.

ukoethe commented:
I tested https://github.com/msarahan/conda/commit/a330766af3c4a24de711ffa117daa1baa9678186 and https://github.com/msarahan/conda-env/commit/bcc603841d843ff4f49d0100c36ecae75094f13c and found the following problems:

1. `python setup.py install` in the conda repo fails with
 ```
   Converting UNIX line endings to Windows
   error: [Error 2] Unable to find file.
```
 The missing file is `unix2dos` which does not normally exist on Windows (not even after `conda install unxutils`). However, on my machine (Windows 7) there is no need to force Windows line endings on .bat-files, Unix line endings work just as well. If you can verify this behavior on all relevant Windows versions (which I think is very likely -- programs requiring Windows line endings have become the exception), the best solution is to just delete line 52 to 63 in `setup.py`.

2. When I do
```
activate foo
activate bar
```
the foo-directories remain in the PATH.

msarahan commented:
@ukoethe thanks.   If that's vestigial, I'm very happy.  I did this at some point when confused by bat files not working.  I'll revert that bit.

2 may be cause by stale activate.bat and deactivate.bat lying around.  Unless you explicitly deleted them, I found it necessary to delete them before doing ```python setup.py develop``` or else the old ones remained.

The quickest way to check this is to look in activate.bat for ..changeps1 (this exists in the old file, but not in the new one.)

I only ask you to check because the unit tests explicitly check this condition (https://github.com/conda/conda/pull/1727/files#diff-d849c4832afeb45c2b7396d369b7d852R182), so I have reasonable confidence that it is working, once the correct activate scripts are in place.  I'm away from a computer for the next few hours, but will double-check it myself when I get connected again.

ukoethe commented:
>  If that's vestigial, I'm very happy.

I make no guarantees :neutral_face:.

> 2 may be caused by stale activate.bat and deactivate.bat lying around.

I supsect that this is the case, but IMHO the install scripts should be able to fix this automatically -- manually deleting `activate.bat` and `deactivate.bat` is probably asking too much from a normal user.

ukoethe commented:
> 2 may be caused by stale activate.bat and deactivate.bat lying around.

For completeness, I manually removed `activate.bat` and `deactivate.bat` and recreated them via `python setup.py install` in `conda-env`, and it recreated the __old__ files (with `..changeps1`). I was a bit suprised about this because your comments say that responsibility for `activate.bat` and `deactivate.bat` was moved from `conda-env` to `conda`. This was apparently not the case in the commits I tried, although I believe that they are the newest available.

msarahan commented:
Sorry for the confusion: as of https://github.com/msarahan/conda-env/commit/bcc603841d843ff4f49d0100c36ecae75094f13c those files are no longer in conda-env.  However, having conda-env and conda both involved in this change, with separate PRs makes this very hard to keep consistent.  Have you:

  - in python install folder (miniconda, or whatever):
      - manually delete activate.bat and deactivate.bat, until I figure out how to get Setuptools or distutils to overwrite them
  - in __conda-env__ git clone:
      - git pull <msarahan> bcc603841d843ff4f49d0100c36ecae75094f13c  
      - python setup.py install
  - in __conda__ git clone:
     -  git pull <msarahan> 93a60971d7089939d17a33b82a1ca8332dba3462
     - python setup.py install

wulmer commented on conda/cli/activate.py:
```diff
@@ -24,95 +38,105 @@ def help():
-        sys.stderr.write("discarding %s from PATH\n" % binpath)
+        path = os.getenv("PATH")
+        if os.getenv("CONDA_ACTIVE_ENV"):
+            binpath = binpath_from_arg(os.getenv('CONDA_ACTIVE_ENV'))
```
I think in those two lines, the os.getenv should have a default of `'root'`. When starting with a new shell, the root environment is implicitly activate, so when activating another env, the root env should be discarded first.

I would suggest to change it into `os.getenv('CONDA_ACTIVE_ENV', 'root')`.

wulmer commented on tests/test_activate.py:
```diff
@@ -1,652 +1,582 @@
             stdout, stderr = run_in(commands, shell)
-            assert stdout == envs + "/test1/bin:" + PATH
-            assert stderr == 'discarding {syspath} from PATH\nprepending {envs}/test1/bin to PATH\n'.format(envs=envs, syspath=syspath)
+            assert_equals(stderr, u'prepending {envpaths} to PATH'\
```
As the root env should be implicitly deactivated, the expected response here should be:
```python
assert_equals(stderr, u'discarding {syspath} from PATH\nprepending {envpaths} to PATH'\
    .format(syspath=pathlist_to_str(_envpaths(root_dir)),
    envpaths=pathlist_to_str(_envpaths(envs, 'test1'))))
```
(first discarding the root env, then activating test1)

msarahan replied:
I see your point, but I don't agree with it.  I don't think the root environment is necessarily on PATH.  I have been working to make that more achievable.  If paths to the activate scripts are provided absolutely, then there is no requirement, explicit or implicit, that root is active when starting a process.  Moreover, my thinking is that assuming no initial root env activation, we're in a better place.  When people want to work with conda and anaconda tools, they can start up a special "anaconda prompt" that drops them into the root environment (or directly into any other environment).  This is how Visual Studio does it.  People are still free to add the root environment to PATH.  If that's the case, then we never touch it - only prepend to PATH so earlier env gets found first.  This is less clean, because stuff not present in the prepended PATH may still get found in the later PATH.  I don't propose to fix this problem, but rather to provide tools towards the (cleaner) alternative.

Does this make sense to you, or do you see a better way?

wulmer replied:
Hm... I see. I don't have any objections against keeping the PATH clean. However there is one important thing:

**There must be a (command line) way to put conda/activate on the PATH.**

In Visual Studio, there is always a `vcars*.bat` or `setenv.cmd` script on the PATH which allows you to quickly add the development environment to the shell you are currently working with. Although there is start menu command to open such a shell in a new terminal, it is *essential* to be able to do that from a script (e.g. Makefile).

For Anaconda, it could work like this:
1. Skip adding `Anaconda\Scripts` and `Anaconda` to the PATH when installing Anaconda.
2. Provide a `anaconda.bat` script in `Anaconda\somepath` which adds `Anaconda\Scripts` and `Anaconda` to the PATH.
3. Put `Anaconda\somepath` on the system PATH

This allows to easily provide a start menu entry (just call `cmd /k anaconda.bat`) but also to integrate Anaconda in custom batch scripts. In my company we have a lot of scripts like the following Jupyter notebook starter (simplified code):
```batch
if not exist "%~dp0venv" (
    conda create -p %~dp0venv --file=requirements.txt jupyter
)
call activate %~dp0venv
jupyter notebook
```
Simply by distributing this batch script together with a requirements file, we can provide the user a single entry point to installing and running Jupyter notebook. If you decide to drop Anaconda from the system path, we have to adapt this code as conda is not on the PATH. The easiest solution would be to add a `call anaconda.bat` at the beginning of the script.

msarahan replied:
This is actually pretty much exactly what I had in mind, and you absolutely are able to call activate.bat from a script (by default now, with no arguments, it activates the root env), but I hadn't thought of a separate folder just for the conda.exe file.  Instead, I had assumed that people would be willing to hard-code the full path to the batch file.  This works well for Microsoft, since Visual Studio is more often installed to a standard location, but our users may get tripped up, since Anaconda and Miniconda are more often installed to user-local paths that vary much more.

I will mull the idea of a folder that has nothing but Conda in it, and adding only that folder to PATH.  I like the idea a lot, but I'm also hesitant to introduce yet another folder to think about.  Do you think this approach extends to Linux/OS X, as well?

wulmer replied:
Typo? **Not** `conda.exe` should be in a separate folder, **but** an `anaconda.bat` script which puts `conda.exe` on the PATH!

I think for Linux, most people have already defined their own shell aliases or command scripts which take care of adding Anaconda to the PATH. Providing them a solution with the installer might be a nice extra. See also [this note](https://groups.google.com/a/continuum.io/d/msg/conda/c0DrrSdy1e8/xooqxrDNy4sJ) from @asmeurer about the Linux integration.

msarahan replied:
Not a typo, more like a brain fart.  Thanks for correcting me.

I really like the idea of a cmd folder with only activate in it, plus the __offer__ of setting up an alias to that activate script.  What I have in mind is:

alias activate "source <full path to activate script>"

Windows would have 3 installation options:
1. Add root env to PATH, as is done now.  python, plus any contained packages will work without additional steps on command line or in scripts.  Leaves room for ambiguity.
2. add proposed cmd folder to path.  Python will not work before activate.bat is called.  Cleaner.  We can also make sure we prepend PYTHONPATH if it is set, and hopefully clear up some newbie issues there.
3. do nothing.  This is still a viable option on windows, because start menu shortcuts to enter an activated environment will be provided.

wulmer replied:
Regarding Windows option 1: When the root env is initially on the PATH, activating another env should remove the root env from the PATH.

msarahan replied:
I think this is hard to solve generally.  The way that I have things now, I consider the whole block of additional paths as one item.  If I need to go find the root env, I have rather large misgivings that users may have changed their PATH themselves, and totally trashed my assumption that I can treat the collection of paths as a single entity for purposes of string replacement.  If that assumption is violated, then searching the path for each piece individually, and re-creating the PATH faithfully afterwards, is a mundane but difficult bookkeeping task.  I could just store the original PATH when any environment is activated, I guess.

This will require quite a bit more work to do right, and I'd like to punt.  I would like to create another issue for this, and solve it soon.  If you have time, I would welcome any effort you can contribute towards this.

wulmer replied:
You are right. Solving this in general seems to be difficult. For our applications, we only have to solve the specific case in which the root env is on the PATH as created by the installer. Any other PATH modifications by the user can be neglected in our case.

I think there are multiple solutions available for this, e.g. when Windows user is choosing option 1. not only put the root env on the PATH but also set CONDA_ACTIVE_ENV to the root env. This way, the active env is first deactivated when another env is activated.

Good idea to follow up on this in a new issue. If you create a new one, I would be happy to contribute.

wulmer commented on tests/test_activate.py:
```diff
@@ -1,652 +1,582 @@
-                stdout, stderr = run_in(commands, shell)
-                assert not stdout
-                assert stderr == 'discarding {syspath} from PATH\nprepending {envs}/test3/bin to PATH\n'.format(envs=envs, syspath=syspath)
+            assert_equals(stderr, u'prepending {envpaths1} to PATH'\
```
Should be
```python
assert_equals(stderr, u'discarding {syspath} from PATH\nprepending {envpaths1} to PATH'\
```
as the root env should first be discarded before the test1 env is being added.

wulmer commented on conda/cli/activate.py:
```diff
@@ -24,95 +38,105 @@ def help():
+        # strip off previous prefix, if any:
+        path = re.sub("^\(.*\)\ ", "", path)
+        env_path = sys.argv[2]
+        if conda.config.changeps1 and env_path:
```
There is a condition branch missing here. Should be:
```python
if conda.config.changeps1:
    if env_path:
        path = "({}) {}".format(os.path.split(env_path)[-1], path)
    else:
        path = "{}".format(path)
```

msarahan replied:
Thanks for pointing this out, but I think it is OK.  The "path" (actually, the prompt text) is first set at https://github.com/conda/conda/pull/1727/files#diff-fc8d33d0af3aba7c6c18a4fddaf6db7aR119 - this is passed in from the call to ..setps1 with the current PS1 (or PROMPT) setting.

If no env_path is set, then PS1 is not changed.

wulmer replied:
> If no en_path is set, then PS1 is not changed.

That's exactly the point why `deactivate.bat` seems to malfunction: In [line 22](https://github.com/conda/conda/pull/1727/files#diff-99f99701021ab30ce1a8e6c0459f4660R22) it calls `call "%CONDA_EXE%" ..setps1 "" "%PROMPT%"` which leads to `env_path == ""` so that the prompt is not changed when deactivating an environment.

Try it out. `activate foo` and `deactivate` still keeps the prompt set to `(foo) $P$G`.

msarahan replied:
I'm very confused by this - it works fine for me:

```
C:\Users\builder>activate py35
prepending C:\Users\builder\MC_Main\envs\py35 and C:\Users\builder\MC_Main\envs\py35\Scripts and C:\Users\builder\MC_Main\envs\py35\Library\bin to PATH

(py35) C:\Users\builder>deactivate
discarding C:\Users\builder\MC_Main\envs\py35 and C:\Users\builder\MC_Main\envs\py35\Scripts and C:\Users\builder\MC_Main\envs\py35\Library\bin from PATH

C:\Users\builder>
```

How can I replicate your system?  Clink?  Cmder?  Do you have any other setup that I can install and test with?

wulmer replied:
Oh yes, sorry. It was a side effect of my Clink problem. When I disable Clink, it works.

wulmer commented on conda/cli/activate.py:
```diff
@@ -24,95 +38,105 @@ def help():
+                if not path:
+                    path = os.getenv("PS1")
+        # strip off previous prefix, if any:
+        path = re.sub("^\(.*\)\ ", "", path)
```
I had a really nasty bug with this line. Although the tests in `test_activate.py` all ran on my system, I could not make it work in my normal Windows cmd shell. This line did **never** strip off the prefix, so that I always ended up with something like `(root) (env1) (root) (foo) C:\what\ever\path>`.

Took a while to figure out the reason: I am using [Clink](http://mridgers.github.io/clink/) which does some magic to the prompt. Even if `%PROMPT%` looks unsuspicious on the command line (`$P$G`), try out the following:
```
>>> import os
>>> os.getenv('PROMPT')
'C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08(bar) C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08C\x08L\x08I\x08N\x08K\x08 \x08$P$G'
```
Wow! This effectively prevents the regexp above from removing the prefix from the prompt as the prompt string doesn't start with `(` even if a prefix is activated. The `%PROMPT%` variable always starts with this garbage.

I solved it for me by changing the regexp to `\(.*\)\ ` (removing the caret) which ignores the first characters. I assume that there are more people out there who are using Clink, so I think somehow a solution must be found for this.

msarahan replied:
argh!  That caret is actually there because zsh expands stuff in parentheses, and I was getting prompts like:

(test) 

on my linux test system.  I will try to come up with a better way - perhaps only replace the first regex match, and ditch the caret?  If people come up with creative ways to colorize the env name, then this code will clearly need more work.

wulmer commented:
Hi @msarahan, thanks for creating this PR, even if *this ended up being a little more work than anticipated*. I really appreciate your efforts.

I have added some line comments about changes I had to make. They will make the tests in `test_activate.py` work.

I had a really nasty bug with Clink, see [my comment on it](https://github.com/conda/conda/pull/1727/files#r43066903).

Keep on going!!!

BTW: Did you know there is a lovely place in India called [Sarahan](https://www.google.de/search?q=sarahan)?

msarahan commented:
@ukoethe here is what I think is necessary to overwrite activate.bat and deactivate.bat.

http://stackoverflow.com/questions/19133831/python-how-to-force-overwriting-of-files-when-using-setup-py-install-distutil

I could also force the copy: ```distutils.dir_util.copy_tree(src, dst)``` (from http://stackoverflow.com/questions/7419665/python-move-and-overwrite-files-and-folders)

I like the first option more.  This is development.  Conda installing conda will take care of this (I'll make absolutely certain of this).  In the meantime, I think it's reasonable to ask developers to add the extra --force flag when testing this PR, or anything dealing with files installed by setuptools/distutils.

Thoughts?

ukoethe commented:
> here is what I think is necessary to overwrite activate.bat and deactivate.bat.

I finally got it to work: one must call `python setup.py develop`. The standard `python setup.py install` did not install activate.bat and deactivate.bat (not even when `--force` was added or `site-packages/conda` was explicitly deleted). Here are some additional remarks:

1. IMHO, the variable `CONDA_ACTIVE_ENV` should hold the full path to the active environment.  Alternatively, one could introduce an additional variable for that purpose. 

2. Why has the variable `CONDA_ENVS_PATH` disappeared?

3. I agree with @wulmer that there should be a (command line) way to put conda/activate on the PATH. The easiest is to create a new directory `<conda_root>\cmd` which holds trivial versions of conda.bat, activate.bat, and deactivate.bat that simply forward the call to the real scripts in `<conda_root>\Scripts`. Then a user can safely keep `<conda_root>\cmd` in the PATH at all times. As a bonus, this eliminates the need to copy the forwarding scripts to every environment. Moreover, a user could replace `<conda_root>\cmd` with another directory holding customized versions of these scripts (e.g. to change the prompt). This wouldn't work if activate.bat etc. were part of every environment (assuming that the environment's `Scripts` directory is at the front of the PATH).

msarahan commented:
> I finally got it to work

Thanks for posting this.  I have found development on Conda to be very painful with a lot of pitfalls.  I hope I'll have time to make this whole process simpler, or at least document the known-working way along with many of the potential pitfalls.

> IMHO, the variable CONDA_ACTIVE_ENV should hold the full path to the active environment. Alternatively, one could introduce an additional variable for that purpose.

Agreed.  This is very important for disambiguation.

> Why has the variable CONDA_ENVS_PATH disappeared?

I think only because it was not present in @wulmer 's PR.  I did not attempt to merge the existing activate.bat with his PR.  I will first verify that conda itself is not looking for this (it might be).  If Conda is not looking for this, I will add it back in.

> 3.

I agree completely, and I'm sorry I didn't respond to this when you first suggested it.  I can't seem to find your other comment now, but I know that you suggested this in an earlier comment.  I think the cmd folder will be a good addition, especially since git has already set the precedent.

msarahan commented:
Regarding CONDA_ENVS_PATH:

The call to binpath_from_arg leads to https://github.com/msarahan/conda/blob/wulmer-fix/activate-behaviour-on-windows/conda/cli/activate.py#L39, which in turn leads to:
https://github.com/conda/conda/blob/wulmer-fix/activate-behaviour-on-windows/conda/cli/common.py#L393

This respects CONDA_ENVS_PATH in conda (https://github.com/msarahan/conda/blob/wulmer-fix/activate-behaviour-on-windows/conda/config.py#L158).  No further additions to activate.bat are necessary.

msarahan commented:
Latest commits move activate.bat and deactivate.bat to cmd folder that @ukoethe and @wulmer proposed.  Work is needed in the installer to change options that add this to PATH.  Also, setup.py is broken right now, since installation of those scripts was done with setup.py.

Fixed:
  - work around shell regex by enclosing our notation in __two__ parentheses.  I could not otherwise come up with a regex that didn't break either zsh (linux/mac) or clink (Windows).  Choosing a different symbol that () might also work.  I'm open to suggestions.
  - removed line ending conversion from setup.py.  Things seem to work fine without it.  All tests still passing.
  - CONDA_ACTIVE_ENV is stored as the full path to the environment prefix.

DO NOT MERGE!  This is broken, until the setup.py has been fixed, and we have alternative ways of installing the cmd folder (and adding it to PATH.)

ispmarin commented:
Any news on this front? Unfortunately this bug is preventing me using conda.

ilanschnell commented on conda/install.py:
```diff
@@ -56,7 +56,7 @@ def __enter__(self):
 
-on_win = bool(sys.platform == 'win32')
+on_win = bool(sys.platform.startswith("win"))
 
```
On Windows (even 64-bit) `sys.platform` is always `win32`.

ukoethe replied:
I know, but why not playing it safe?

JanSchulz replied:
IMO the safe variant is checking for `win32` because if that is ever changed, probably a few assumptions are also changed and this might result in code you wanted to guard with this. (ok, this is probably a really unrealistic scenario given some stories about MS not abandoning any windows API even if it is 30 years old and actually a bug...)

msarahan replied:
I don't really think this is worth arguing over.  I'm happy to revert it if need be.

ilanschnell commented:
What is the reason for renaming `CONDA_DEFAULT_ENV`?  Doing so will break backwards compatibility for many scripts people have written.

JanSchulz commented on conda/config.py:
```diff
@@ -171,7 +171,7 @@ def pkgs_dir_from_envs_dir(envs_dir):
 # ----- default environment prefix -----
 
-_default_env = os.getenv('CONDA_DEFAULT_ENV')
+_default_env = os.getenv('CONDA_ACTIVE_ENV')
```
This looks wrong... `_default_env != CONDA_ACTIVE_ENV`

msarahan replied:
Incomplete rename; probably will be reverted.  Thanks for pointing it out, regardless.

JanSchulz commented:
As someone who has one local script which checks for `CONDA_DEFAULT_ENV` and a few appveyor configs which do, I would also be very unhappy if that env variable is renamed...

msarahan commented:
@ilanschnell the reason for renaming CONDA_DEFAULT_ENV is because it is a confusing name.  Please refer to discussion in this PR for more detail.  However, I agree that this change is not really related to the spirit of this PR, and would be happy to revert that change and propose it elsewhere.

@JanSchulz I agree - I hate changing API's.  I also hate when new people get stuck due to poor nomenclature.  It's a difficult situation, but I lean towards making it easier on newbies.  For everyone else, it is an annoyance, followed by a "search-and-replace".

ukoethe commented:
I'm in favor of `CONDA_ACTIVE_ENV`, as the old name is too confusing. @JanSchulz: If you are against the Win32 API, you can't seriously argue in favor of `CONDA_DEFAULT_ENV` :-)

JanSchulz commented:
Hey, I also find `CONDA_DEFAULT_ENV` confusing, but I've seen a few breaking API changes which resulted in bad things / public relation afterwards...

ilanschnell commented:
As an historical note, the reason the variable was named `CONDA_DEFAULT_ENV` is because commands such as `conda list`, `conda install`, `conda remove`, etc. have an option `-n` and `-p`, which allow you to be explicit about the conda environment, but they *default* to `CONDA_DEFAULT_ENV`.  I am personally not a fan of using the word "active" or "deactive" as it implied that you have to "activate" an environment before programs in it can be used, which is not the case.

msarahan commented:
Notes from this latest effort:

- round-trip from Win PATH to MSYS/cygwin path is unreliable
- MSYS/cygwin PATH is differently ordered from windows PATH.  Needs its own "reference" state
- case sensitivity for string comparison has become an issue with going back and forth.  String comparisons now all do lower() before comparing.
- track reference PATH and PS1 states on each activation.  This approach is global state, but only in a per-shell basis.  Is this OK?  It's a lot simpler than asking conda for the deactivated PATH for some reason.

wulmer commented:
@msarahan Thanks for resuming this issue. I really would like to support you and start with testing your latest version. How can I set up my local Miniconda installation to use your branch version of conda?

msarahan commented:
@wulmer thanks!  Much appreciated.  Some setup steps here, but this may not be complete, just due to the not-quite-always-reproducible nature of build environments (I may have something installed that I've forgotten to list):

- tests of bash assume that bash is somewhere on PATH.  This probably means that "git" needs to be a test-time requirement of this package, at least on windows, which packages "Git for Windows" which includes bash.  What this means for you is that you'll need to install some form of bash somewhere on PATH that stays on PATH through the test process (anywhere not in the conda root env should be fine).  This PR is meant to support bash generally on Windows, so I'm curious to test the different ways of getting it: cygwin, MSYS, MSYS2 - please add any others you're aware of, so that we minimize blind spots.

- Testing conda almost always ruins my miniconda install at some point.  I recommend a dedicated miniconda install, without adding it to PATH (or, to make life easier, a different user or a clean VM and add it to PATH).  If you don't add the installation, to PATH, make sure all of your exe calls use the full path to the exe, just to avoid confusion.  From there, ```conda install setuptools pytest``` (because they is missing otherwise), then remove the conda folder and conda egg-info from site-packages.  Finally, ```<full path to dedicated miniconda python> setup.py develop``` in your git checkout.  Note that ```develop``` doesn't keep shell scripts up to date - only Python files.  You may need to re-run it, or re-run ```<python> setup.py install``` if ```develop``` is not copying shell scripts appropriately (but this might be a good bug to fix, so please report it or fix and PR)

- Run tests with full path to py.test.exe.  Mine is something like ``` c:\Miniconda2\Scripts\py.test.exe Z:\Documents\code\conda\tests\test_activate.py``` - z: is my network drive out of my VM.  Adjust accordingly for your system.

- The cmd.exe test suite runs OK, I believe.  The bash test suite on windows is about 2/3 there.  I have not yet tried this stuff on Linux or Mac.

Finally, I would like to move towards supporting different shells as simply config files, rather than things being hard-coded into Conda, for the sake of modularity and ease of adding other shells.  There is a glimpse of this in the test_activate file, where the shell collection is now a dictionary.  This is something I'm saving for a future PR, so that this PR doesn't get delayed any longer.  I welcome your thoughts and comments, though, especially if anything in this PR can make that future work easier.

msarahan commented:
Testing multiple shells was a huge pain with the loop structure.  I replaced it with py.test parametrization.

To run specific tests, do something like this:

    <full path to py.test.exe> --shell="cmd.exe" tests/test_activate.py

or for more than one shell:

    <full path to py.test.exe> --shell="cmd.exe" --shell="msys" tests/test_activate.py

or for a specific test with a specific shell:

    <full path to py.test.exe> --shell="cmd.exe" tests/test_activate.py::test_activate_test1[cmd.exe]

That last one could probably be improved.  Presently, the default shell setting is [] (no tests at all, --shell required).  The defaults could be set to a list of string shell names.  However, when I've set that as the default, it seems that passing --shell does not override it, and I get all shells run.  Seems to be related to https://bugs.python.org/issue16399

wulmer commented on conda/cli/activate.py:
```diff
@@ -24,95 +52,114 @@ def help():
+    Format a path list, e.g., of bin paths to be added or removed,
+    for user-friendly output.
+    """
+    return ' and '.join([path.replace("\\\\", "\\") for path in paths])
```
Assuming that your path list contains the normalized paths, this `replace` is not necessary. And it might accidentally hide some internal string inconsistencies.

wulmer commented on conda/install.py:
```diff
@@ -86,6 +88,36 @@ def win_soft_link(src, dst):
+        if src.endswith("conda"):
+            src = src + ".exe"
+
+        p = subprocess.check_call(["bash", "-l", "-c", "ln -sf {src} {dst}".format(
```
Should be checked first if `bash` is available on the `PATH` or at least be caught.

msarahan replied:
thanks.  I have had that in here, but removed it for testing so that I see any errors.  I have had a lot of trouble getting output from py.test - it seems the -s flag doesn't always work somehow.  I'll re-do the exception handling you suggest, as it is definitely the right way to do things.

wulmer commented on conda/utils.py:
```diff
@@ -75,6 +77,47 @@ def url_path(path):
+def translate_stream(stream, translator):
+    translated_stream = ""
+    for line in stream.split("\n"):
+        for term in shlex.split(line):
```
Sorry, wrote my previous comment into the commit of this line.
For reference: This line breaks `set PATH` output on Windows as `shlex.split(r'C:\Miniconda') => ['C:Miniconda']`

wulmer commented on tests/helpers.py:
```diff
@@ -44,11 +37,11 @@ def run_conda_command(*args):
-    stdout, stderr = p.communicate()
-    return (stdout.decode('utf-8').replace('\r\n', '\n'),
-        stderr.decode('utf-8').replace('\r\n', '\n'))
+    stdout, stderr = [stream.strip().decode('utf-8').replace('\r\n', '\n').replace('\\\\', '\\')
```
Why should you `replace` double backslashes here? Where do they come from? Which conda command line output produces double backslashes?

wulmer commented on tests/test_activate.py:
```diff
@@ -1,74 +1,194 @@
+            printpath="echo %PATH%",
+            raw_ps=os.getenv("PROMPT", ""),
+            exe="cmd.exe",
+            shell_args="/d /c",
```
As you are later using `shell_args.split(" ")`, it might be better to store a list instead here: `shell_args=['/d', '/c']`. This prevents possible errors in the future of the kind `shell_args='/a /b /c "C:\\Program Files"'`...

msarahan replied:
yep, that's a good improvement.  Thanks.

wulmer commented on tests/test_activate.py:
```diff
@@ -1,74 +1,194 @@
+        cmd_bits = [shells[shell]["exe"]] + shells[shell]["shell_args"].split(" ") + [cmd_script.name]
+        p = subprocess.Popen(cmd_bits, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = p.communicate()
+        os.unlink(cmd_script.name)
```
To make sure that the `cmd_script` is cleaned up correctly, I would recommend a `try finally`.

msarahan replied:
good call.

wulmer commented on tests/test_activate.py:
```diff
@@ -1,74 +1,194 @@
+        p = subprocess.Popen(cmd_bits, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = p.communicate()
+    output_translator = shells[shell]["path_from"]
+    streams = [translate_stream(stream.strip().decode('utf-8').replace('\r\n', '\n').replace('\\\\', '\\'), output_translator)
```
Same comment as [before](https://github.com/conda/conda/pull/1727/files#r51996742): Why do you have to `replace('\\\\', '\\')` here? What produces these double backslashes?

wulmer commented on tests/test_activate.py:
```diff
@@ -1,74 +1,194 @@
+
+def _envpaths(env_root, env_name=""):
+    if 'win' in platform:
+        paths = [join(env_root, env_name).rstrip("\\"),
```
I don't think the `rstrip` is required here: The [documentation of `os.path`](https://docs.python.org/2/library/os.path.html#os.path.join) states that there will only be a separator at the end if the last part is empty.

msarahan replied:
I think it is, because when env_name is "" (the default) - then you end up with the separator at the end.  This gets called a few places with the default:  ```pathsep.join(_envpaths(root_dir))```

wulmer replied:
Thanks, I overlooked the obvious, sorry. Perhaps a comment could clarify that for the future.

wulmer commented on tests/test_activate.py:
```diff
@@ -76,577 +196,476 @@ def _write_entry_points(envs):
-            source {deactivate}
-            printf $PATH
-            """).format(envs=envs, deactivate=deactivate)
+def test_activate_deactivate():
```
Missing `shell` argument, should be `def test_activate_deactivate(shell):`

msarahan replied:
Good catch, thanks.

msarahan commented:
@wulmer thanks for all the great feedback.  The present mess with backslashes is my confusion over when they are escape characters, and when they are literals.  When passing anything to shell subprocesses, I think we need 4 (first set of escapes are Python, next set of escapes are shell).  When we print stuff, there should be only 2 (resulting in just one being displayed).

It is exceedingly painful to test this stuff with the string matching we do. Much of what is a perfectly valid result is marked incorrect because there's too many slashes, or the slashes are going the wrong way.  I'm hoping the "translate_stream" function can help simplify that.

wulmer commented:
Hm.... yeah, backslashes are always a great mess. However escaping them (`\\\\`) should not be necessary at all:
* In `cmd.exe`, you don't have to escape `\` as it is no meta character there.
* In other shells, e.g. Cygwin's `bash` on Windows, you might have to escape Windows paths, for example when converting paths using `cygpath`, but `cygpath` also understands normal slashes instead of backslashes. Simply write your Windows paths with normal slashes there. So there is no need to use backslashes at all in other shells than `cmd.exe`.
* IMHO, the output of conda commands should *never* contain escaped backslashes. Whatever takes the output of a conda command should take care of handling the output correctly, e.g. when passing back to a shell again, but the output should always be the same, independent of who is calling.

electronwill commented:
When this PR is merged in, the docs team should revisit https://github.com/conda/conda-docs/pull/248 to update the test drive as noted by @msarahan .

ilanschnell commented on conda/install.py:
```diff
@@ -56,7 +56,9 @@ def __enter__(self):
             pass
 
-on_win = bool(sys.platform == 'win32')
+from conda.utils import win_path_to_unix
```
Please see the comment at the top of this module.  This module needs to be standalone.

msarahan replied:
Thanks for pointing this out.  I have taken a similar approach that you did: try to import it, but fall back to defining it.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/79bb6d16-4548-4c55-907c-aab6996b428a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/79bb6d16-4548-4c55-907c-aab6996b428a) completed: **1 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/0c8ba9d9-4493-4d46-b02b-d63cc703183c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/0c8ba9d9-4493-4d46-b02b-d63cc703183c) completed: **1 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/b1796ad1-ad03-420d-857b-dfb7079b53c3) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/b1796ad1-ad03-420d-857b-dfb7079b53c3) completed: **1 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/68a284ff-18dd-4033-865e-3ff8d45c7952) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/68a284ff-18dd-4033-865e-3ff8d45c7952) completed: **1 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/8ece05d6-e7ac-43a1-8a7a-c48d45bb3439) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/ecc8038f-2ad5-4145-b05b-04deb8536a62) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/ecc8038f-2ad5-4145-b05b-04deb8536a62) completed: **2 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/2923bcee-ac26-48a8-a544-61624546136c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/2923bcee-ac26-48a8-a544-61624546136c) completed: **2 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/656ed618-6614-4a57-9580-7fd4fae620e4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/656ed618-6614-4a57-9580-7fd4fae620e4) completed: **2 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/79851bad-5df4-48b7-9739-a69bfa91d6e0) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/79851bad-5df4-48b7-9739-a69bfa91d6e0) completed: **2 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/1fb8a557-9259-481e-9493-39e90896e6a4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/1fb8a557-9259-481e-9493-39e90896e6a4) completed: **2 new issues, 36 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/5df6e64d-c165-4c51-b7bf-ab2e3fa84543) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/5df6e64d-c165-4c51-b7bf-ab2e3fa84543) completed: **2 new issues, 37 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/44a34284-47c3-40ee-936a-fcb24405bca9) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/44a34284-47c3-40ee-936a-fcb24405bca9) completed: **13 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/931e4c44-b554-4081-8342-4994e37d76ab) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/931e4c44-b554-4081-8342-4994e37d76ab) completed: **13 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/e8f9f3d3-aec5-4d80-b081-959df389a0c0) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/e8f9f3d3-aec5-4d80-b081-959df389a0c0) completed: **14 updated code elements**

msarahan commented:
Note for anyone testing this, and for packagers:

  - Installing without --old-and-unmanageable puts the activate/deactivate scripts into an egg, where they are not found.
  - development-mode installations do not seem to copy the cmd folder at all.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/aeb372be-ebaa-4654-a6ee-7863e7fc6ade) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/aeb372be-ebaa-4654-a6ee-7863e7fc6ade) completed: **14 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/4e12e662-7a1a-4109-834e-ff217907f87f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/4e12e662-7a1a-4109-834e-ff217907f87f) completed: **13 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/0e783937-1735-4978-9049-35cd51780450) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/0e783937-1735-4978-9049-35cd51780450) completed: **13 updated code elements**

msarahan commented:
@ilanschnell @wulmer @ukoethe @JanSchulz @kalefranz @MaxNoe 

This is finally ready for some polish.  The tests are passing on all platforms, now including Windows with Git Bash.  I'm open to improvements and suggestions.  Specifically:

  - rename of CONDA_DEFAULT_ENV to CONDA_ACTIVE_ENV - can we get a vote or something on this?  @kalefranz ultimately this is your call.
  - I have not integrated fish shell tests here.  I intend to do that in a future PR closely following this one.  It needs to be soon, because activate scripts should live in cmd now, and the current arrangement has the fish script in bin.
  - We need to update our installers to have the cmd folder on PATH when PATH is modified.
  - We need to update our installers with the option to have ONLY the cmd folder on PATH.  This will be the least disruptive to other installed software.

Are there any test cases for activation/deactivation that I have missed?  Note that I have added test cases for:

  - spaces in path
  - relative path
  - (currently known fail) non-ascii characters in path

if there are other aspects you use in your workflow, please chime in.  I'll be working tomorrow to verify that the contents of CONDA_ENVS_PATH is taken into account.  This is handled in Conda, so it should work, but I don't think we have explicit tests for it.

ilanschnell commented:
I have so many questions about this PR, I don't know where to start.  Here are a few:
  * What is the motivation behind renaming `CONDA_DEFAULT_ENV`?
  * Why is `psutil` now a dependency?
  * What are the Windows to Unix path conversion functions used for?
  * How exactly does the symlinking of conda into environments change?
  * How did this PR get so large to being with?
  * Why wasn't it possible to create a few smaller (and more manageable) PRs instead?

Thanks for all the hard work.

ukoethe commented:
I tested this PR on Windows 7 and could NOT get it to work. I installed a fresh miniconda from `Miniconda3-latest-Windows-x86_64.exe` into `c:\conda-test` and then did the following in a cmd.exe shell:
```
set PATH=c:\conda-test\Scripts;%PATH%
conda install conda-build pytest
cd <HOME>\conda-git
git checkout wulmer-fix/activate-behaviour-on-windows # latest commit: 32fc0d366
c:\conda-test\python setup.py develop
c:\conda-test\python setup.py install
```
This does not create the directory `c:\conda-test\cmd`. `activate.bat` and `deactivate.bat` still reside in `c:\conda-test\Scripts`. Accordingly, `conda build ...` gives the error message:
```
ln: creating symbolic link `/C/conda-test/envs/_build/cmd/activate' to 
     `/C/conda-test/cmd/activate': No such file or directory
```
The test suite doesn't work either:
```
cd <HOME>\conda-git
c:\conda-test\Scripts\py.test.exe tests\test_activate.py
    py.test-script.py: error: unrecognized arguments: --cov-report tests\test_activate.py
```

msarahan commented:
@ukoethe a few pointers:

1. setup.py install does not put the cmd folder in the right place unless you also tell it not to put stuff into an egg.  I have been using:

    python setup.py install --old-and-unmanageable

2.  You need to run the tests in the conda root folder, and be sure that you have installed pytest-cov.  Alternatively, you can remove or rename setup.cfg to not run coverage.

3.  To run test_activate.py you need to specify shells.  here's an example for windows (assumes that you have installed git bash with ```conda install git```):

    py.test tests\test_activate.py --shell=cmd.exe --shell=git_bash

More details in this comment: https://github.com/conda/conda/pull/1727#issuecomment-179923341

msarahan commented:
@ilanschnell 

> What is the motivation behind renaming CONDA_DEFAULT_ENV?

Never mind, I misspoke.  I renamed things back to CONDA_DEFAULT_ENV in https://github.com/msarahan/conda/commit/c94063cc33e0f77d49e34463a12cd1f7faca26c2

> Why is psutil now a dependency?

psutil is necessary to determine what the parent shell is.  This is in turn necessary to make conda behave appropriately (slash behavior and help output) for shells other than cmd.exe on Windows.

> What are the Windows to Unix path conversion functions used for?

Primarily for test_activate.py, but they see some use in symlinking also.  The main reason why is that os.path.join always joins with platform-native slashes, but these are inappropriate for bash shells on Windows.

> How exactly does the symlinking of conda into environments change?

On Windows, bat files are created in environments that call back to the activate scripts in the root environment.  Bash on windows uses ln in a subprocess call.  Powershell is currently not implemented.  Linux/Mac have not changed.

> How did this PR get so large to being with?

This PR is more than a year old.  It was originally proposed as https://github.com/conda/conda/pull/913/ in September 2014.  It required fixing after the conda-conda env split. @wulmer proposed https://github.com/conda/conda/pull/1707, and I created this PR to fix his code a bit.  At that point, Aaron had left, and had left us with no Windows CI, because he tied Appveyor to his personal account with Appveyor.  This PR thus included some changes to fix Appveyor, because this PR made drastic changes to Windows.  (those have now been removed, because Kale fixed them independently of this PR)  We had the Anaconda 2.4.0 release, and this PR got lost.  Then @MaxNoe submitted his change to add bash support on windows.  I invited him to submit it here, because it really should depend/build on the work that @wulmer did.  If @wulmer's work had been merged earlier, then the bash stuff could have been separate.  The "bash stuff" includes all of the path translation and psutil addition, as well as fixes to conda required to fix the invalid assumption that when sys.platform==win32, then windows paths are always what you're dealing with.

> Why wasn't it possible to create a few smaller (and more manageable) PRs instead?

This question is perhaps better answered with another question: why wasn't it possible to merge this PR in any of its earlier iterations?  This PR has snowballed because it hasn't been merged, and subsequent development (such as the bash support on Windows) depends on the changes made earlier.  PRs that depend on PRs are difficult to manage at best.

ukoethe commented:
> python setup.py install --old-and-unmanageable

not a very encouraging option name, but worked :wink:

> be sure that you have installed pytest-cov

There is no conda package for pytest-cov, I had to install via pip

> py.test tests\test_activate.py --shell=cmd.exe

output (corresponding [junit.xml](https://github.com/conda/conda/files/154890/junit.xml.txt)):
```
platform win32 -- Python 3.5.1, pytest-2.8.5, py-1.4.31, pluggy-0.3.1
rootdir: <HOME>\conda, inifile: setup.cfg
plugins: cov-2.2.1
collected 19 items
tests\test_activate.py ..................x
```
> py.test tests\test_activate.py --shell=git_bash

This unsurprisingly failed because I didn't install bash via conda. I think the option should be renamed into `conda_bash` to make clear that it refers to the path `<conda_root>\Library\bin\bash.exe`.

I edited `test_activate.py` to add an option `bash` which assumes that `bash.exe` is already in the PATH. Subsequent `py.test tests\test_activate.py --shell=bash` worked with one expected failure as above.

msarahan commented:
@ukoethe Thanks for the feedback.

A test that assumes bash is on PATH is better, definitely.  My only reservation about that is Cygwin, which needs to be special-cased somehow, since it treats paths with the /cygdrive prefix.

the ```--old-and-unmanageable``` flag is equivalent to ```--single-version-externally-managed --record=record.txt```

I prefer it for brevity, but it is awkward to read.

ukoethe commented:
I'm not sure if this is related to the present PR: `conda build ...` still fails for the following reason: When conda-build sets up the _build environment, ito copies `c:\conda-test\Scripts\conda.exe` into `c:\conda-test\envs\_build\conda` (without extension!). Later, `fix_staged_scripts()` thinks this file is a unix script and tries to convert it into a `.bat` script. Of course, that cannot work. Why is conda.exe copied in the first place, and why without extension?

msarahan commented:
That is absolutely related to this PR.  I think that's a bug in the symlinking logic, which is done to keep conda and the activate scripts on PATH when the root environment is replaced with the activated environment.  I will investigate.  Thanks for reporting.

ilanschnell commented:
@msarahan thanks for your responses.

If `psutil` is only used on Windows to determine the parent shell, then `psutil` should only be a dependency on Windows.

msarahan commented:
I think the capability to know the parent shell will be valuable in Linux/Mac also.  For example, this will probably be useful in integrating Fish, since it is fairly different from bash/zsh.  I'd like to leave it in.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/85c25f65-7737-49ca-8df9-84e8c005ea3b) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/85c25f65-7737-49ca-8df9-84e8c005ea3b) completed: **16 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/73f5fe20-f7cf-4eaa-919b-5044f60a486a) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/fae5a137-70a3-40ad-b607-dc3748c42682) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/ab4df8d9-845f-4ee1-9de9-61537a4f654f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/conda/conda/inspections/ab4df8d9-845f-4ee1-9de9-61537a4f654f) completed: **11 updated code elements**

kalefranz commented:
Want to rebase here and merge to master?  I think we're ready to go...

kalefranz commented on conda/install.py:
```diff
@@ -56,7 +57,21 @@ def __enter__(self):
+try:
+    from conda.utils import win_path_to_unix
+except ImportError:
+    def win_path_to_unix(path, root_prefix=""):
```
This appears to be duplicated in `conda/utils.py`.

kalefranz replied:
Oh, I understand why.  Never mind.

kalefranz commented on conda/utils.py:
```diff
@@ -75,6 +79,43 @@ def url_path(path):
 
 
+def win_path_to_unix(path, root_prefix=""):
+    """Convert a path or ;-separated string of paths into a unix representation
```
This appears to be duplicated in `conda/install.py`. I'd prefer to have it here.

msarahan replied:
I would too.  It's duplicated because of https://github.com/conda/conda/pull/1727#discussion_r54364354

kalefranz commented on setup.py:
```diff
@@ -34,6 +34,12 @@
+                                    "cmd/activate.bat", "cmd/deactivate.bat"]), ]
+            }
+else:
+    kwds = {'data_files':  [("cmd", ["cmd/activate", "cmd/deactivate"]), ]}
```
@ilanschnell doesn't like the `cmd` prefix.  I agree with you that they're not quite bin files either.  Still, I'd be more comfortable with something that's more FHS-esque.

msarahan replied:
This is done specifically to mimic the behavior of git for windows.  See https://github.com/conda/conda/pull/1727#discussion_r42775466

Please suggest something else if you'd like to break that convention.

kalefranz commented:
Sorry @msarahan, going to have to merge master again.  I was hoping you'd get this in earlier today, but went ahead and broke your PR.

What I merged to master is a `conda.common` sub-package that is a "bottom layer" to separate imports.  Anything within shouldn't import *any* other conda module outside of `conda.common`.  The big difference for this PR I think, besides fixing some imports, is moving the `conda/utils.py` code to `conda/common/utils.py`.

msarahan commented:
Ok. That's fine. I'll try to find time soon.  I seem to have messed up unrelated tests somehow - probably incorrect conflict resolution.

kalefranz commented:
Do you want to do the honors of clicking that big green merge button?

kalefranz commented:
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

vmuriart commented:
does this address #626 too?

vmuriart commented:
Just got around to testing, it didn't. Was #913 not meant to add that functionality for powershell?
